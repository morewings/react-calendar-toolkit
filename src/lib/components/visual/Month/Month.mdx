---
name: Month
menu: Customize UI
---
import { Playground, Props } from 'docz';
import {Datepicker} from 'lib';
import Month from './Month';

`Month` component displays specific date with [__month precision__](/#calendar-views-and-precision). `onDateSet` callback prop allows to set provided Date object as selected in `Datepicker`. Replace it with custom component.

<Playground>
  {() => {
    const style = {
      boxSizing: 'border-box',
      flex: '1 1 calc(100% / 3 - 4px)',
      color: 'black',
      lineHeight: '50px',
      textAlign: 'center',
      padding: '0',
      border: '2px dashed pink',
      margin: '2px',
    };
    const CustomComponent = ({
     onDateSet,
     date,
     isDisabled,
     isSameMonth,
     isSelected,
     name,
     isHighlighted,
   }) => {
      const handleClick = () => {
        console.log('date set', date);
        onDateSet(date);
      };
      return (
        <div onClick={handleClick} style={style}>
          {name.abbreviated}
        </div>
      )
    };
    return (
      <Datepicker
        onDateSet={date => {console.log('date set', date);}}
        minPrecision="month"
        renderMonthAs={CustomComponent}
        title="Demo datepicker" />
    )
  }}
</Playground>

### Props

<Props of={Month} />


## Wrap with custom CSS class

You can override CSS classname of Day Calendar element, see `monthCalendarClassName` prop.

```jsx
<Datepicker
  onDateSet={date => {console.log('date set', date);}}
  minPrecision="day"
  monthCalendarClassName="your-custom-className"
  title="Demo datepicker" />
```
