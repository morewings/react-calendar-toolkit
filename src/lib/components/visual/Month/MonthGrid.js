import React from 'react';
import PropTypes from 'prop-types';
import classes from './Month.module.css';

const MonthGrid = ({children, wrapperClassName}) => (
  <div className={wrapperClassName} role="grid">
    {children}
  </div>
);

MonthGrid.propTypes = {
  wrapperClassName: PropTypes.string,
  children: PropTypes.node.isRequired,
};

MonthGrid.defaultProps = {
  wrapperClassName: classes.grid,
};

export default MonthGrid;
