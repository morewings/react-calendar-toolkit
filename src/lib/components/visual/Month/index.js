export {default} from './Month';
export {default as MonthGrid} from './MonthGrid';
