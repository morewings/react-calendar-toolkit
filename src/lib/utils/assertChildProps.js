import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
