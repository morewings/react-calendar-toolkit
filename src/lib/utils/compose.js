const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
