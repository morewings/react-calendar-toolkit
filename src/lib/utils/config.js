export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
