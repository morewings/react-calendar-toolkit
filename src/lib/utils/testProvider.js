import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
