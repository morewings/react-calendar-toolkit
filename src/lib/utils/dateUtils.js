import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
