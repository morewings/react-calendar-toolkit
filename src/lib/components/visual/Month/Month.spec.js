import React from 'react';
import {DatepickerMockProvider as Provider} from 'lib/utils/testProvider';
import {render, fireEvent} from '@testing-library/react';
import Month from './Month';

const defaultProps = {
  onDateSet: jest.fn(),
  date: new Date(2020, 0, 12),
  name: {
    numeric: 1,
    wide: 'wide',
    abbreviated: 'abbreviated',
    narrow: 'narrow',
  },
};

describe('Month', () => {
  beforeEach(() => {
    defaultProps.onDateSet.mockClear();
  });

  describe.each`
    belongsToSameMonth | isCurrent | isSelected | isDisabled | isHighlighted | isWeekend
    ${true}            | ${true}   | ${true}    | ${true}    | ${true}       | ${true}
    ${false}           | ${true}   | ${true}    | ${true}    | ${true}       | ${true}
    ${false}           | ${false}  | ${false}   | ${false}   | ${false}      | ${false}
  `(
    'display props',
    ({
      belongsToSameMonth,
      isCurrent,
      isSelected,
      isDisabled,
      isHighlighted,
      isWeekend,
    }) => {
      it(`it renders`, () => {
        const props = {
          ...defaultProps,
          belongsToSameMonth,
          isCurrent,
          isSelected,
          isDisabled,
          isHighlighted,
          isWeekend,
        };
        const {asFragment} = render(<Month {...props} />, {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        });
        expect(asFragment()).toMatchSnapshot();
      });
    }
  );

  it('handles click when belongsToSameMonth === true', () => {
    const props = {
      ...defaultProps,
      belongsToSameMonth: true,
      isCurrent: false,
      isSelected: false,
      isDisabled: false,
      isHighlighted: false,
      isWeekend: false,
    };
    const {getByRole} = render(<Month {...props} />, {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    fireEvent.click(getByRole('button'));
    expect(defaultProps.onDateSet).toHaveBeenCalledTimes(1);
    expect(defaultProps.onDateSet).toHaveBeenCalledWith(defaultProps.date);
  });
});
