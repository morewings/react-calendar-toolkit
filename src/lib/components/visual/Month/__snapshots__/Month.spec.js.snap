// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month display props it renders 1`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    aria-label="wide"
    class="wrapper isSelected isDisabled isCurrent isHighlighted"
    role="button"
    tabindex="-1"
  >
    abbreviated
  </div>
</DocumentFragment>
`;

exports[`Month display props it renders 2`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    aria-label="wide"
    class="wrapper isSelected isDisabled isCurrent isHighlighted"
    role="button"
    tabindex="-1"
  >
    abbreviated
  </div>
</DocumentFragment>
`;

exports[`Month display props it renders 3`] = `
<DocumentFragment>
  <div
    aria-disabled="false"
    aria-label="wide"
    class="wrapper"
    role="button"
    tabindex="0"
  >
    abbreviated
  </div>
</DocumentFragment>
`;
